Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,172 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Metrics API open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the Swift Metrics API project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Swift Metrics API project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Combine
+import Foundation
+
+@available(OSX 10.15, *)
+public enum SystemMetricsProvider {
+    fileprivate static let lock = ReadWriteLock()
+    fileprivate static let queue = DispatchQueue(label: "com.apple.CoreMetrics.SystemMetricsProvider", qos: .background)
+    fileprivate static var timeInterval: DispatchTimeInterval = .seconds(2)
+    fileprivate static var cancellable: Cancellable?
+    fileprivate static var systemMetricsType: SystemMetrics.Type = UnsuportedMetricsProvider.self
+
+    public static func bootstrapSystemMetrics(pollInterval interval: DispatchTimeInterval = .seconds(2), systemMetricsType: SystemMetrics.Type? = nil) {
+        self.lock.withWriterLockVoid {
+            self.timeInterval = interval
+            if let type = systemMetricsType {
+                self.systemMetricsType = type
+            } else {
+                #if os(Linux)
+                self.systemMetricsType = LinuxSystemMetricsProvider.self
+                #else
+                self.systemMetricsType = UnsuportedMetricsProvider.self
+                #endif
+            }
+        }
+        self.updateSystemMetrics()
+    }
+
+    public static func cancelSystemMetrics() {
+        self.cancellable?.cancel()
+    }
+
+    fileprivate static func updateSystemMetrics() {
+        let interval = self.lock.withReaderLock { self.timeInterval }
+        let c = self.queue.schedule(after: .init(.now()), interval: .init(interval)) {
+            let prefix = "process_"
+            let pid = ProcessInfo.processInfo.processIdentifier
```

Comment:
Should we read and store the pid just once rather than many times, is this lookup cheap?